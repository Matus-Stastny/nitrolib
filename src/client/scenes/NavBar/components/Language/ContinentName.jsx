// @flow strict
import * as React from "react";

import { tKeys } from "client/records/Continents";
import Text from "client/components/Text";

type Props = {|
  id: string,
|};

const ContinentName = (props: Props) => <Text t={tKeys[props.id]} />;

export default ContinentName;
