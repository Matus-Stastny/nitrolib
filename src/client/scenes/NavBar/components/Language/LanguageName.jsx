// @flow strict
import * as React from "react";
import styled from "styled-components";

import config from "client/consts/config";
import type { Language } from "client/records/Languages";
import LanguageNameText from "./LanguageNameText";

type Props = {|
  language: Language,
|};

const Icon = styled.i`
  background: ${(props: { flagId: string }) =>
    `url("${config.imagesUrl}flags/32x32/${props.flagId}.png")`};
  height: 32px;
  width: 32px;
  display: block;
  transform: scale(0.7);
`;

const LanguageName = ({ language }: Props) => (
  <>
    <Icon flagId={language.flag} />
    <LanguageNameText>{language.name}</LanguageNameText>
  </>
);

export default LanguageName;
