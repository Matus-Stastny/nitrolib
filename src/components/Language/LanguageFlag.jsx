// @flow strict
import styled from "styled-components";
import PropTypes from "prop-types";

import config from "../../consts/config";

type Props = {
  flagId: string,
  scale: number,
};

const LanguageFlag = styled.i`
  background: ${({ flagId }: Props) => `url("${config.imagesUrl}flags/24x0/flag-${flagId}.jpg")`};
  height: 13px;
  border-radius: 2px;
  width: 24px;
  margin-right: 8px;
  xdisplay: block;
`;

// $FlowIssue
LanguageFlag.propTypes = {
  flagId: PropTypes.string.isRequired,
};

LanguageFlag.defaultProps = {
  scale: 1,
};

export default LanguageFlag;
