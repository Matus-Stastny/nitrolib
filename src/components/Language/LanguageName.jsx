// @flow strict
import * as React from "react";
import styled from "styled-components";

import type { Language } from "../../records/Languages";
import config from "../../consts/config";
import LanguageNameText from "./LanguageNameText";

type Props = {|
  language: Language,
|};

const Icon = styled.i`
  background: ${({ flagId }: { flagId: string }) =>
    `url("${config.imagesUrl}flags/32x32/${flagId}.png")`};
  height: 32px;
  width: 32px;
  display: block;
  transform: scale(0.7);
`;

const LanguageName = ({ language }: Props) => (
  <>
    <Icon flagId={language.flag} />
    <LanguageNameText>{language.name}</LanguageNameText>
  </>
);

export default LanguageName;
