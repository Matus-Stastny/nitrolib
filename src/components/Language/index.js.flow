// @flow strict
import * as React from "react";

type Props = {|
  onChange: (lang: string) => void,
|};

declare export default React.ComponentType<Props>;
