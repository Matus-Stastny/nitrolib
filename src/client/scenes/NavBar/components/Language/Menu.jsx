// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled from "styled-components";

import * as fetchedContext from "client/services/fetched/context";
import Flex from "client/primitives/Flex";
import Text from "client/components/Text";
import mq from "client/services/utils/mediaQuery";
import type { ThemeProps } from "client/records/Brand";
import type { Languages, Language } from "client/records/Languages";
import LanguageName from "./LanguageName";
import ContinentName from "./ContinentName";

// used for decide, when to render 2/3 columns
// and when to activate customHeight on Popup
const LANGUAGES_LIMIT = 5;
// width and height of language rows in px
const LANGUAGE_ROW_HEIGHT = 31;
const LANGUAGE_ROW_WIDTH = 180;

const MenuWrapper = styled.div`
  background-color: white;
  position: absolute;
  top: 45px;
  right: 20px;
  left: auto;
  display: flex;
  border-radius: 3px;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, 0.1);
  ${mq.ltTablet`
    left: 20px;
  `};
`;

const ContinentList = styled.div`
  min-width: 200px;
  font-size: 12px;
  font-weight: 500;
  ${mq.ltTablet`
    display: none;
  `};
`;

type ActiveProps = ThemeProps & {|
  active: boolean,
|};

const ContinentItem = styled(Flex)`
  background-color: ${(props: ActiveProps) => props.active && props.theme.colors["neutral-100"]};
  color: ${(props: ActiveProps) => props.active && props.theme.colors["primary-600"]};
  padding: 14px 16px;
  line-height: 16px;
  cursor: pointer;
  &:hover {
    background-color: ${(props: ActiveProps) => props.theme.colors["neutral-100"]};
    color: ${(props: ActiveProps) => props.theme.colors["primary-600"]};
  }
`;

const LanguageList = styled.div`
  padding: 10px;
  line-height: 20px;
  box-sizing: border-box;
  overflow-y: auto;
  overflow-x: hidden;
  max-height: calc(100vh - 60px);
  ${mq.gtTablet`
    width: 560px;
  `};
  ${mq.ltTablet`
    padding: 15px 12px;
  `};
`;

type SizeProps = {|
  width: number,
  height: number,
|};

const LanguageListWrapper = styled.div`
  width: ${(props: SizeProps) => props.width};
  height: ${(props: SizeProps) => props.height};
  max-height: 682px;
  ${mq.gtTablet`
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
  `};
`;

const LanguageItem = styled.div`
  background-color: ${(props: ActiveProps) => props.active && props.theme.colors["primary-600"]};
  color: ${(props: ActiveProps) => props.active && props.theme.colors.white};
  width: 33%;
  height: 26px;
  display: inline-flex;
  align-items: center;
  border-radius: 3px;
  padding-left: 5px;
  margin-bottom: 5px;
  box-sizing: border-box;
  cursor: pointer;
  &:hover {
    background-color: ${(props: ActiveProps) => props.theme.colors["neutral-200"]};
  }
  ${mq.ltTablet`
    width: 100%;
  `};
`;

type State = {|
  continent: string,
|};

export default class Menu extends React.Component<{}, State> {
  state = {
    continent: "",
  };

  getWidthConstant = (filteredLanguages: Language[]) =>
    filteredLanguages.length < LANGUAGES_LIMIT ? 2 : 3; // 2 || 3 columns

  getLanguageWrapperHeight = (filteredLanguages: Language[]) => {
    const widthConstant = this.getWidthConstant(filteredLanguages);

    // For filtered cases - enables to wrap to (2 || 3) cols every time
    const heightConstraint =
      Math.ceil(filteredLanguages.length / widthConstant) * LANGUAGE_ROW_HEIGHT; // custom height depends on number of columns
    return filteredLanguages.length <= LANGUAGES_LIMIT // && this.props.isMobile
      ? filteredLanguages.length * LANGUAGE_ROW_HEIGHT // custom height depends on number of languages, only on mobile && when less languages than limit
      : heightConstraint;
  };

  getLanguageWrapperWidth = (filteredLanguages: Language[]) =>
    this.getWidthConstant(filteredLanguages) * LANGUAGE_ROW_WIDTH;

  handleContinent = (continent: string) => {
    this.setState({ continent });
  };

  changeLanguage = (languageId: string) => {
    const loc = window.location;
    const parts = loc.pathname.split("/");
    const restOfUrl = parts.slice(2).join("/");

    window.location.assign(`${loc.origin}/${languageId}/${restOfUrl}${loc.search}`);
  };

  filterLanguages = (languages: Languages) =>
    R.values(languages).filter(
      language => this.state.continent === "" || this.state.continent === language.continent,
    );

  render = () => (
    <fetchedContext.Consumer>
      {fetched => {
        const filteredLanguages = this.filterLanguages(fetched.languagesData.languages);
        return (
          <MenuWrapper>
            <ContinentList>
              <ContinentItem onClick={() => this.handleContinent("")}>
                <Text t={__("common.languages_all")} />
              </ContinentItem>
              {fetched.languagesData.continents.map(continent => (
                <ContinentItem
                  onClick={() => this.handleContinent(continent.id)}
                  key={continent.id}
                  active={continent.id === this.state.continent}
                >
                  <ContinentName id={continent.id} />
                </ContinentItem>
              ))}
            </ContinentList>
            <LanguageList>
              <LanguageListWrapper
                height={this.getLanguageWrapperHeight(filteredLanguages)}
                width={this.getLanguageWrapperWidth(filteredLanguages)}
              >
                {filteredLanguages.map(language => (
                  <LanguageItem key={language.id} onClick={() => this.changeLanguage(language.id)}>
                    <LanguageName language={language} key={language.id} />
                  </LanguageItem>
                ))}
              </LanguageListWrapper>
            </LanguageList>
          </MenuWrapper>
        );
      }}
    </fetchedContext.Consumer>
  );
}
