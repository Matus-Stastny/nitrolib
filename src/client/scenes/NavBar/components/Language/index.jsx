// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled from "styled-components";

import * as intlContext from "client/services/intl/context";
import * as fetchedContext from "client/services/fetched/context";
import type { ThemeProps } from "client/records/Brand";
import ClickOutside from "client/components/ClickOutside/index";
import LanguageName from "./LanguageName";
import Menu from "./Menu";
import LanguageNameText from "./LanguageNameText";

type State = {|
  shown: boolean,
|};

const OpenButton = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;
  &:hover {
    ${LanguageNameText} {
      color: ${(props: ThemeProps) => props.theme.colors["primary-600"]};
    }
  }
`;

export default class Language extends React.Component<{}, State> {
  state = {
    shown: false,
  };

  handleToggle = () => {
    this.setState({ shown: !this.state.shown });
  };

  render() {
    return (
      <fetchedContext.Consumer>
        {fetched => (
          <intlContext.Consumer>
            {intl =>
              R.has(intl.language.id, fetched.languagesData.languages) && (
                <>
                  <OpenButton onClick={this.handleToggle}>
                    <LanguageName language={fetched.languagesData.languages[intl.language.id]} />
                  </OpenButton>
                  {this.state.shown && (
                    <ClickOutside onClickOutside={this.handleToggle}>
                      <Menu />
                    </ClickOutside>
                  )}
                </>
              )
            }
          </intlContext.Consumer>
        )}
      </fetchedContext.Consumer>
    );
  }
}
