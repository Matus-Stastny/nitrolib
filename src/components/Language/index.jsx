// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled, { css } from "styled-components";

import MobileSelect from "./MobileSelect";
import ClickOutside from "../ClickOutside";
import * as intlContext from "../../services/intl/context";
import * as fetchedContext from "../../services/fetched/context";
import { themeDefault } from "../../records/Theme";
import type { ThemeProps } from "../../records/Theme";
import LanguageName from "./LanguageName";
import Menu from "./Menu";
import LanguageNameText from "./LanguageNameText";
import mq from "../../styles/mediaQuery";

const OpenButton = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkDark};
  cursor: pointer;
  &:hover {
    ${LanguageNameText} {
      color: ${({ theme }: ThemeProps) => theme.orbit.paletteProductNormal};
    }
  }
  ${mq.ltTablet(css`
    display: none;
  `)};
`;

OpenButton.defaultProps = {
  theme: themeDefault,
};

const Wrapper = styled.span`
  display: flex;
  flex-direction: row;

  ${mq.gtTablet(css`
    display: none;
  `)};
`;

OpenButton.defaultProps = {
  theme: themeDefault,
};

type Props = {|
  onChange: (lang: string) => void,
  flat: boolean,
|};

type State = {|
  shown: boolean,
|};

export default class Language extends React.Component<Props, State> {
  static defaultProps = {
    flat: false,
  };

  state = {
    shown: false,
  };

  handleToggle = () => {
    this.setState(state => ({ shown: !state.shown }));
  };

  render() {
    const { onChange, flat } = this.props;
    const { shown } = this.state;

    return (
      <fetchedContext.Consumer>
        {fetched => (
          <intlContext.Consumer>
            {intl =>
              R.has(intl.language.id, fetched.brandLanguage.languages) && (
                <>
                  <OpenButton onClick={this.handleToggle}>
                    <LanguageName language={intl.language} />
                  </OpenButton>
                  <Wrapper>
                    <MobileSelect
                      current={intl.language}
                      languages={R.values(fetched.brandLanguage.languages)}
                      onChange={onChange}
                    />
                  </Wrapper>
                  {shown && (
                    <ClickOutside onClickOutside={this.handleToggle}>
                      <Menu onChange={onChange} flat={flat} />
                    </ClickOutside>
                  )}
                </>
              )
            }
          </intlContext.Consumer>
        )}
      </fetchedContext.Consumer>
    );
  }
}
