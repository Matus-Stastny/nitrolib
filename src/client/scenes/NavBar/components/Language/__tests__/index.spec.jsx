// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Language from "../index";

describe("#Language", () => {
  test("render", () => {
    const wrapper = shallow(<Language />);

    expect(wrapper).toMatchSnapshot();
  });

  test("opens when clicked", () => {
    const wrapper = shallow(<Language />);
    wrapper.instance().handleToggle();

    expect(wrapper.state().shown).toEqual(true);
  });
});
