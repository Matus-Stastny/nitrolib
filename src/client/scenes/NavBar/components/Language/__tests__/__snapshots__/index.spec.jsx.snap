// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Language render 1`] = `<[object Object] />`;
